src/graph/transform/cpu/knn.cc

```diff
@@ -15,6 +20,221 @@ namespace dgl {
 namespace transform {
 namespace impl {
 
+// This value is directly from pynndescent
+static constexpr int NN_DESCENT_BLOCK_SIZE = 16384;
+
+/*!
+ * \brief Compute Euclidean distance between two vectors, return positive
+ *  infinite value if the intermediate distance is greater than the worst
+ *  distance.
+ */
+template <typename FloatType, typename IdType>
+FloatType EuclideanDistWithCheck(const FloatType* vec1, const FloatType* vec2, int64_t dim,
```

Comment:
How do we choose between these two Euclidean Distance functions?

Reply:
If we do not have the `worst_distance`, for example when we initialize the distance buffer, we can use `EuclideanDist` without check. If we have the `worst_distance`, we can use the `worst_distance` as an upper bound for optimization (if current distance > worst distance, stop computation and return `inf`)